### DISCUSSION TITLE
Why `from_orm` calls database again

### DISCUSSION THREAD

**Discussion by dsharma522** [pydantic/pydantic#6197] [Question] [answered]
I am using sqlalchemy and pydantic in fast api application. I have orm queries using join. When using `from_orm` from pydantic , it is calling database again. I confirmed via enabling sqlalchemy logs. 

### Original problem

```python
notification_rules_subquery = (
    db.query(NotificationRuleModel.id).filter_by(**rule_filter).subquery()
)
rows = (
    db.query(NotificationRuleModel, NotificationModel)
    .join(NotificationModel, NotificationModel.rule_id == NotificationRuleModel.id)
    .filter(... Some filter)
    .all()
)
db.commit()
```

```python

# Sqlalchemy db model

ModelBase = declarative_base()

@declarative_mixin
class TimestampMixin:
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(DateTime, nullable=False, server_default=func.now(), onupdate=func.now())


class NotificationRule(ModelBase):
    __tablename__ = _NOTIFICATION_RULE_TABLE_NAME
    id = Column(Integer, unique=True, primary_key=True)
    community_id = Column(UUID(as_uuid=True), nullable=False)

class Notification(ModelBase, TimestampMixin):
    __tablename__ = _NOTIFICATION_TABLE_NAME

    id = Column(Integer, primary_key=True, unique=True, autoincrement=True)
    uuid = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_id = Column(Integer, ForeignKey(NotificationRule.id), nullable=False)

# pydantic model
class NotificationRule(BaseModel):
    id: int
    community_id: uuid.UUID
    
    class Config:
        orm_mode = True

class NotificationBaseResponse(DateMixin):
    uuid: uuid.UUID
    id: int
    rule_id: int
    time_frame_start: Optional[datetime]

    class Config:
        orm_mode = True


```

The rows returned by the above query is the list of tuples in form (NotificationRule, Notification), so
When doing 
```
rule = NotificationRule.from_orm(rows[0][0])
```

this calls database again. I am unable to figure out why.

### Another case

I debugged it and found that even on single table query, the from_orm is calling database

```python
row = (
        db.query(NotificationModel)
        .filter(NotificationModel.rule_id == some_id)
        .order_by(NotificationModel.created_at.desc())
        .limit(1)
        .all()
    )
    db.commit()
```

When calling on `Notification.from_orm(row)` on above, it is calling database again. 
However, If i change the query as 
```
row = (
        db.query(*inspect(NotificationModel).attrs)
        .filter(NotificationModel.rule_id == some_id)
        .order_by(NotificationModel.created_at.desc())
        .limit(1)
        .all()
    )
    db.commit()
```
(added *inspect(NotificationModel).attrs )

and then calling from_orm on it, It does not call database again.

### Logs for ref of join query
```bash
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|BEGIN (implicit)
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|SELECT notification_rule.id AS notification_rule_id, notification_rule.community_id AS notification_rule_community_id, notification_rule.user_id AS notification_rule_user_id, notification_rule.purpose AS notification_rule_purpose, notification_rule.targets AS notification_rule_targets, notification_rule.frequency AS notification_rule_frequency, notification_rule.send_at AS notification_rule_send_at, notification_rule.rule AS notification_rule_rule, notification_rule.active AS notification_rule_active, notification_rule.created_at AS notification_rule_created_at, notification_rule.last_run_at AS notification_rule_last_run_at, notification_rule.scheduled_next_run AS notification_rule_scheduled_next_run, notification_rule.status AS notification_rule_status, notification_rule.task_id AS notification_rule_task_id, notification.created_at AS notification_created_at, notification.updated_at AS notification_updated_at, notification.id AS notification_id, notification.uuid AS notification_uuid, notification.rule_id AS notification_rule_id_1, notification.status AS notification_status, notification.time_frame_start AS notification_time_frame_start
FROM notification_rule JOIN notification ON notification_rule.id = notification.rule_id
WHERE notification.status IN (%(status_1_1)s, %(status_1_2)s) AND notification_rule.id IN (SELECT anon_1.id
FROM (SELECT notification_rule.id AS id
FROM notification_rule
WHERE notification_rule.community_id = %(community_id_1)s) AS anon_1) ORDER BY notification.id DESC
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|[generated in 0.00036s] {'community_id_1': UUID('74b46d04-e8c5-4ab9-a466-9721b4bd8e4c'), 'status_1_1': <NotificationStatus.SENT: 'SENT'>, 'status_1_2': <NotificationStatus.DELIVERED: 'DELIVERED'>}
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|COMMIT
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|BEGIN (implicit)
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|SELECT notification_rule.id AS notification_rule_id, notification_rule.community_id AS notification_rule_community_id, notification_rule.user_id AS notification_rule_user_id, notification_rule.purpose AS notification_rule_purpose, notification_rule.targets AS notification_rule_targets, notification_rule.frequency AS notification_rule_frequency, notification_rule.send_at AS notification_rule_send_at, notification_rule.rule AS notification_rule_rule, notification_rule.active AS notification_rule_active, notification_rule.created_at AS notification_rule_created_at, notification_rule.last_run_at AS notification_rule_last_run_at, notification_rule.scheduled_next_run AS notification_rule_scheduled_next_run, notification_rule.status AS notification_rule_status, notification_rule.task_id AS notification_rule_task_id
FROM notification_rule
WHERE notification_rule.id = %(pk_1)s
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|[generated in 0.00012s] {'pk_1': 2}
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|SELECT notification.created_at AS notification_created_at, notification.updated_at AS notification_updated_at, notification.id AS notification_id, notification.uuid AS notification_uuid, notification.rule_id AS notification_rule_id, notification.status AS notification_status, notification.time_frame_start AS notification_time_frame_start
FROM notification
WHERE notification.id = %(pk_1)s AND notification.uuid = %(pk_2)s
INFO:     2023-06-21 09:49:36|base.py -> _log_info[136]|sqlalchemy.engine.Engine|[generated in 0.00013s] {'pk_1': 5, 'pk_2': UUID('e1b6020d-57c2-4bbc-8c3c-99cdfe3e3a32')}
```

**Comment by connebs** [comment, marked as answer]
This is a bit confusing to read through since you seem to have changed a bunch of names around inconsistently to sanitize your code for this question. Generally it is much easier for others to debug such an issue if you just make a simple and complete repro that others can run.

However, in this particular case my assumption is that this is just an ORM behaving as ORM's do. Lazy-loading is a very common pattern, where things are not actually queried from the database until you request the particular object/property in question.

e.g. if you have a `Notification` which has a relationship to a `NotificationRule` and can be accessed like this: `Notification.rule`, the ORM (SQLAlchemy in this case) won't load that `.rule` until you specifically ask for it. And of course when loading a model using `.from_orm()`, Pydantic will be asking for it, causing a database query to be triggered so that SQLAlchemy can get that `NotificationRule` so that Pydantic can use it.

This pattern in particular is why N+1 queries can be a common footgun  when using an ORM (and why ORMs frequently get a bad rep). 

Again, hard to say what you should do in particular since it seems like the models you've posted are a bit incomplete, but this is mostly likely something you want to tackle on the SQLAlchemy side using its eager loading functionality.

[This blog post might be able to help.](https://sgoel.dev/posts/handling-the-n-1-selects-problem-in-sqlalchemy/)

**Reply by samuelcolvin** [reply]
Great answer.

**Reply by jordantshaw** [reply]
@samuelcolvin  @acnebs 
This is a great answer describing a very common problem.  However I don't see a great solution on how to manage un-wanted lazy loads.  For example, if you use an ORM to run a specific query, and then you want to load that result into a pydantic model.  If you have a relationship property and that property is defined in the pydantic model, it would be lazy loaded upon model validation.

However this can be very frustrating, because maybe that property was left out of the ORM query for a reason, and loading that relationship is not wanted.  Is there anyway in pydantic to avoid lazy loading specific fields?  Like some kind of "exlcude" parameter or something.

**Reply by connebs** [reply]
Why would you have that property in the pydantic model if you don't want it? If you want the property, use the eager loading features of the ORM. If you don't want the property, don't include it in the pydantic model.

**Reply by jordantshaw** [reply]
I think this is a fairly common use case.  For example, you may have a specific database property you may or may not want loaded, depending on user input. If the property is not loaded, then in the pydantic model, if that property is unset, you may have a default value you want to display.  The problem is that default value will never get displayed because the pydantic model will trigger a lazy load upon model validation. 

There are ways around this.  The first possible solution that doesnt really work, is you can set that property in the ORM to not load at all, based off user input.  The problbem with this approach is most ORMs will now return that value as None, rather than leaving it unset.  So the pydantic model will see that the property is None, instead of unset, and the default values will not be displayed.

The other option, which seems like a huge workaround  is to dump the ORM result object to a python dictionary, before the pydantic model validation. But now you have to worry about how to get the ORM result object into a python dictionary correctly.  This is clunky, and kind of defeats the purpose of being able to let pydantic handle the validation and serialization. It would be a much better experience if pydantic supported some kind of partial load or exclude option on model validation.
